extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.kafka.streams.runtime.graal;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.rocksdb.CompressionType;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.RecomputeFieldValue.Kind;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+/**
+ * loadLibrary() loads the RocksDB native libraries and is called from multiple static initializers.
+ * It's designed to run only once (guarded by the {@code libraryLoaded} field). As I don't know how
+ * to reset this field to its initial value (cannot access the original {@code LibraryState} type
+ * from within the substitution class, the entire method is substituted.
+ */
+@TargetClass(className = "org.rocksdb.RocksDB")
+final class Target_org_rocksdb_RocksDB {
+
+    private enum LibraryState {
+        NOT_LOADED,
+        LOADING,
+        LOADED
+    }
+
+    @Alias
+    @RecomputeFieldValue(kind = Kind.FromAlias)
+    private static AtomicReference<LibraryState> libraryLoaded = new AtomicReference<>(LibraryState.NOT_LOADED);
```

Comment:
@Sanne, this is the field I'd like to reset. I tried to substitute the class as you suggested, but it seems I can't as it is an enum type:

    com.oracle.svm.core.util.UserError$UserException: Annotated class must inherit directly from Object: class io.quarkus.kafka.streams.runtime.graal.Target_org_rocksdb_RocksDB$Target_org_rocksdb_RocksDB_LibraryState

So subsituting the entire method that makes use of the field is the only way I've found so far. If there's a better/easier one, that'd be great.

Reply:
hey @gunnarmorling , Substitutions are indeed a bit limited. They are also a bitch to maintain in the long term, so don't look at them via your perfectionist hat, but war the pragmatic one; personally I've used them to speed-up things as otherwise one ends up in the neverending loop of waiting for 3rd parties to accept PRs and wait for all such releases to align.

So replace all the method and don't feel too bad about it :)

Either way, the proper solution is to craft a PR for the upstream project, so to not live with such stuff for too long. Sometimes that's not entirely possible, but you could e.g. refactor the upstream project to minimize the scope of substitutions.

e.g. PostgreSQL JDBC no longer needs substitutions , the ones we have for Hibernate ORM are much smaller/simpler now than what they used to be (and should reduce further with further ORM work), the ones for Narayana are about to become no longer necessary either, etc...

Reply:
P.S. to answer your question: there might be a better way, but I don't know it :) Either way, suggesting to not bother.

Reply:
Thanks, @Sanne! So one alternative would be to delay initialization of all the classes calling this method in the static initializer to runtime. This would avoid the substitution here mostly. I was originally shying away from it, but it might be better indeed. What's the "Quarkus way" of doing this?

Reply:
it depends on your needs. Generally we try to "run at build" if that would imply either better dead-code-elimination possibilities (e.g. the constant being resolved having some effect like disabling a feature) or if it would otherwise speed-up the runtime bootstrap (by avoiding doing some work at runtime).

If you have neither such benefits, might be better to do such things at runtime. Especially if you have this code need to re-run at runtime, then there's no need to bother running it during the build.

Reply:
Yes, so how do I configure a class to be initialized at runtime with Quarkus then?

Reply:
Never mind, I've found what I need in the Substrate build items.